### ISSUE TITLE
Key tables for copy modes

### ISSUE THREAD

**Issue by eraserhd** [tmux/tmux#1953]
### Issue description

I'm building a plugin that adds a key table, `copy-mode-kakoune`, which I activate after switching to copy mode.

Key tables are normally one-shot.  To remedy this, each mapping in the table calls `switch-client -Tcopy-mode-kakoune`.

Then, bumping an unmapped key leaves the mode, which is pretty bad UX.  As a result, I iterate through all keys and map all of them to stay in my new copy mode.  This is ugly and brittle (I don't handle mouse mappings, and tmux can add keys in the future, and figuring out which keys are available required reading a bunch of code), but it sort of works, but then...

After doing this, my root bindings no longer work.

#### Make the copy-mode key table an option

Add a new option that allows overriding of the copy-mode table.  If set, return that value from `window_copy_key_table`; otherwise, fall back to the existing logic of inspecting `mode-keys` and returning one of the two existing tables.

I think this would prevent it from being one-shot.

#### Add a "default" binding

Add a way to make a binding for a table that handles anything unbound.  This way, I can make one binding that stays in the mode, and bumping unmapped keys doesn't lose the mode.  I can still exit the mode normally by binding some key that calls `send-keys -X cancel` and doesn't call `switch-client -Tcopy-mode-kakoune`.

This is where this approach gets tricky, I think: we want the default binding to have lower precedence than root bindings, so the code that calls the copy mode bindings will need to look  up keys in two passes.  The first to find a positive mapping, and if that fails the second to find the first default mapping.

#### Allow configuring of tables to be "sticky"

In other words, some tables are not one shot.  Perhaps an option to `bind-key`?  This might get weird because I'm not sure what happens if I do `select-keys -X cancel`, I think the table would still be active, so there would need to be a mechanism to unstick the table.

#### Continue to Slog Away

I could iterate over the root table and reinstall those mappings in my new table if they don't conflict with ones I want in the copy mode.

#### Others

I think there could be other solutions here.

### Required information

```
tmux next-3.1
Linux unknown
tmux-256color / xterm-256color
```

**Comment by eraserhd** [comment]
Er, and I would do the work of course.

**Comment by nicm** [comment]
You could use the Any key binding?

**Comment by eraserhd** [comment]
@nicm Oh, nice.  I double-checked, and this is not documented.  Should I add it to the man page?

This still blocks root bindings, but otherwise makes the code much, much nicer.

I've discovered another manifestation of my primary issue... If I enter my custom copy mode, then use prefix key and come back to the pane, I'm back in the _normal_ copy mode.  Whatever handles the prefix key (which is apparently not the root table?) doesn't preserve the table.

I've thought about running a script on Any that executes root bindings, but don't have a way to get the key that was pressed (as far as I can tell), and the idea of patching my table with root bindings, which should work ... aside from losing the custom copy mode after.

Maybe I should just clear and repopulate the vi or emacs tables?  This would definitely work.

**Comment by eraserhd** [comment]
(verified that it does work)

**Comment by nicm** [comment]
Yes the Any key should be added to the manual, that would be good.
